driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -67,7 +69,9 @@
     private final AtomicBoolean isClosed = new AtomicBoolean();
     private final AtomicBoolean opened = new AtomicBoolean();
     private final Semaphore writing = new Semaphore(1);
-    private final Semaphore reading = new Semaphore(1);
+    private final Lock readerLock = new ReentrantLock(false);
```

Comment:
We don't need a semaphore here.  It's just a lock. And I had to make it re-entrant because otherwise the tests with mock streams will deadlock.  I'll point out where the deadlock would be below.